app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/partner/PartnerNameController.scala

```diff
@@ -0,0 +1,229 @@
+                    partner.partnerType match {
+                      case Some(SCOTTISH_PARTNERSHIP) =>
+                        getPartnershipRedirectUrl(appConfig.scottishPartnershipJourneyUrl,
+                                                  appConfig.partnerGrsCallbackUrl(existingPartnerId)
+                        ).map(journeyStartUrl => SeeOther(journeyStartUrl).addingToSession())
+                      case Some(GENERAL_PARTNERSHIP) =>
+                        getPartnershipRedirectUrl(appConfig.generalPartnershipJourneyUrl,
+                                                  appConfig.partnerGrsCallbackUrl(existingPartnerId)
+                        ).map(journeyStartUrl => SeeOther(journeyStartUrl).addingToSession())
+                      case _ =>
+                        //TODO later CHARITABLE_INCORPORATED_ORGANISATION & OVERSEAS_COMPANY_NO_UK_BRANCH will have their own not supported page
+                        Future(
+                          Redirect(
+                            organisationRoutes.OrganisationTypeNotSupportedController.onPageLoad()
+                          )
+                        )
+                    }
+                  case _ =>
+                    Future.successful(Redirect(dropoutCall))
+                }
+              case Left(error) => throw error
+            }
+        )
+    else
+      throw new IllegalStateException("Partner type does not permit user supplied names")
+
+  private def updateInflightPartner(
+    formData: PartnerName
+  )(implicit req: JourneyRequest[AnyContent]): Future[Either[ServiceError, Registration]] =
+    update { registration =>
+      registration.inflightPartner.map { partner: Partner =>
+        registration.withInflightPartner(Some(setPartnershipNameFor(partner, formData)))
+      }.getOrElse {
+        registration
+      }
+    }
+
+  private def updateExistingPartner(formData: PartnerName, partnerId: String)(implicit
+    req: JourneyRequest[AnyContent]
+  ): Future[Either[ServiceError, Registration]] =
+    update { registration =>
+      registration.withUpdatedPartner(partnerId,
+                                      partner => setPartnershipNameFor(partner, formData)
+      )
+    }
+
+  private def setPartnershipNameFor(partner: Partner, formData: PartnerName): Partner = {
+    val partnershipDetailsWithPartnershipName = {
+      partner.partnerPartnershipDetails match {
+        case Some(partnerPartnershipDetails) =>
+          Some(partnerPartnershipDetails.copy(partnershipName = Some(formData.value)))
+        case None =>
+          // Partnership detail has not been created yet; we need to create a minimal one to carry the user supplied name
+          // until the GRS callback can fully populate it
+          partner.partnerType.map { partnerType =>
+            // This does not look like a good fit; the optionally on one of these can probably be relaxed or tightened.
+            PartnerPartnershipDetails(partnershipType = partnerType,
```

Comment:
`partnershipType` is being removed in the near future, right?

Reply:
Yeah... this block of code highlights the potential removal; we'll get it on a a separate PR.